Sources/CacheAdvance/CacheReader.swift

```diff
@@ -64,6 +65,11 @@ final class CacheReader {
                 // If the previous read was also empty, then the file has been corrupted.
                 throw CacheAdvanceError.fileCorrupted
             }
+            if seekForwardOnly && startingOffset > FileHeader.expectedEndOfHeaderInFile {
+                // If seekForwardOnly is true, then return nil when startingOffset is ahead
+                // Otherwise, there will be an infinite loop in reading messages.
```

Comment:
I'm not understanding how `seekForwardOnly` doesn't lead to data loss when reading a file where `shouldOverwriteOldMessages` is `true`.

If `shouldOverwriteOldMessages` is `true`, then we might have started reading messages at an offset in the file after `FileHeader.expectedEndOfHeaderInFile`. If `seekForwardOnly` is `true`, then we'll never read the messages between `FileHeader.expectedEndOfHeaderInFile` and `offsetInFileAtEndOfNewestMessage`, right?

Data loss is problematic. I think it'd be preferable to detect that we're in an infinite loop and abort rather than asking customers of this tool to expect data loss in certain circumstances.

Reply:
@dfed  I check the this line of [code](https://github.com/dfed/CacheAdvance/blob/main/Sources/CacheAdvance/CacheAdvance.swift#L123), and find the writer begins after `FileHeader.expectedEndOfHeaderInFile` when `shouldOverwriteOldMessages` is true. It is the same logic for `shouldOverwriteOldMessages is false `, the writer begins after `FileHeader.expectedEndOfHeaderInFile` too.

When `shouldOverwriteOldMessages` is true, the file handle should not seek back either. Otherwise, it will run into an infinite loop as well.

Reply:
When set both `shouldOverwriteOldMessages` and `seekForwardOnly` true, it won't cause data loss for normal cache file.
The `seekForwardOnly` flag is used avoid the reader to run into this loop: 

Sequence of normal cache file reading:
1. file header
2. message 1 header
3. message 1 body
4. message 2 header
5. message 2 body
6. …
7. end of file

Sequence of bad cache file ( offsetInFileAtEndOfNewestMessage incorrect ) reading:
1. file header
2. message 1 header
3. message 1 body
6. got empty data, then goto step  1)
7. loop forever

Reply:
In [the line of code you linked](https://github.com/dfed/CacheAdvance/blob/0e63e448850a0d64a2a228d59d293749c3c1a248/Sources/CacheAdvance/CacheAdvance.swift#L123), we're seeking back to `FileHeader.expectedEndOfHeaderInFile` to write new messages, but older messages still start later in the file. When we open this file to read it, we'll start reading at [`header.offsetInFileOfOldestMessage`](https://github.com/dfed/CacheAdvance/blob/0e63e448850a0d64a2a228d59d293749c3c1a248/Sources/CacheAdvance/CacheAdvance.swift#L196). When `seekForwardOnly` is `true`, we'll never read newer messages between `FileHeader.expectedEndOfHeaderInFile` and `header.offsetInFileAtEndOfNewestMessage`.

Note that:
* `header.offsetInFileOfOldestMessage` is always greater than or equal to `FileHeader.expectedEndOfHeaderInFile`
* `header.offsetInFileOfOldestMessage ` may be greater than `header.offsetInFileAtEndOfNewestMessage` if messages have been overwritten

I think we could be more cautious when reading and `shouldOverwriteOldMessages` is **false** (which if I recall correctly is how Airbnb uses CacheAdvance for logging). When a file doesn't overwrite new messages the file handle should _not_ seek back when reading since there should never be newer data at the beginning of that file. If that non-overwriting file had been corrupted, then [this guard](https://github.com/dfed/CacheAdvance/blob/0e63e448850a0d64a2a228d59d293749c3c1a248/Sources/CacheAdvance/CacheReader.swift#L48) might not catch the expected end-of-file, and then we could infinite loop.

I think we can patch the "cache with `shouldOverwriteOldMessages == false` is corrupted results in infinite loop when reading messages" problem I just described by throwing `CacheAdvanceError.fileCorrupted` when we encounter an `.emptyRead` on a cache that should not overwrite old messages. I believe we can make this change without introducing new API.

Reply:
It may _also_ be possible to end up in an infinite loop when reading a corrupted cache where `shouldOverwriteOldMessages == true`. But I think we can detect that case + throw `CacheAdvanceError.fileCorrupted` without introducing new API – we can likely do this by keeping track of whether the reader has already seeked to the beginning of the file and if so, throwing if we're seeking back to the beginning of the file again.

That said, I'm also not convinced we need to solve both corrupted cache cases in the same PR.

Reply:
@dfed  the context provided are really useful. Put together with the atomic issue, we should more clear about the logics.

For this part, I think `offsetInFileAtEndOfNewestMessage` should be always greater if the file is not  corrupted.

> `header.offsetInFileOfOldestMessage` may be greater than `header.offsetInFileAtEndOfNewestMessage` if messages have been overwritten

I read the [code](https://github.com/dfed/CacheAdvance/blob/main/Sources/CacheAdvance/CacheAdvance.swift#L115) when `header.overwritesOldMessages` = true.
If `cacheHasSpaceForNewMessageBeforeEndOfFile` false, the writer will truncate the cache file and continue to append message. In this case, `offsetInFileAtEndOfNewestMessage` should be greater. And if I`cacheHasSpaceForNewMessageBeforeEndOfFile` is true, the writer continue to append message, `offsetInFileAtEndOfNewestMessage` should be greater too.

Reply:
I'd like to have a new parameter for safety, and this could enable me to have an experiment to see the treatment group in Airbnb App. 
> I believe we can make this change without introducing new API.

Reply:
> I think we could be more cautious when reading and shouldOverwriteOldMessages is false (which if I recall correctly is how Airbnb uses CacheAdvance for logging).

yes, it is false in Airbnb app.

Reply:
> If you want to ship a version of this repo with the Airbnb app where you can turn this new functionality off, I recommend shipping a fork of this repo.

Emm, that's a better way to do. Really thanks for your careful review and comments!

Reply:
I see there is a line that resets the `offsetInFileOfOldestMessage` value. And then it is synced to the header.
```
                // We know the oldest message is at the beginning of the file, since we are about to toss out the rest of the file.
                reader.offsetInFileOfOldestMessage = FileHeader.expectedEndOfHeaderInFile
                try reader.seekToBeginningOfOldestMessage()

...
            // Update the offsetInFileOfOldestMessage in our header before we write the message.
            // If the application crashes between writing the header and writing the message data, we'll have lost the messages between the previous offsetInFileOfOldestMessage and the new offsetInFileOfOldestMessage.
            try header.updateOffsetInFileOfOldestMessage(to: offsetInFileOfOldestMessage)

```
Do I have the right understanding about this line?

Reply:
You're right that on this line we set the `reader`'s `offsetInFileOfOldestMessage` to `expectedEndOfHeaderInFile`:
https://github.com/dfed/CacheAdvance/blob/0e63e448850a0d64a2a228d59d293749c3c1a248/Sources/CacheAdvance/CacheAdvance.swift#L126-L127

However, the intermediary lines covered by `...` are key here. On this line (part of your first `...`) we seek the reader forward to beginning of the next message _beyond_ the `expectedEndOfHeaderInFile`.
https://github.com/dfed/CacheAdvance/blob/0e63e448850a0d64a2a228d59d293749c3c1a248/Sources/CacheAdvance/CacheAdvance.swift#L129-L130

Then (part of your second `...`) we `prepareReaderForWriting`:
https://github.com/dfed/CacheAdvance/blob/0e63e448850a0d64a2a228d59d293749c3c1a248/Sources/CacheAdvance/CacheAdvance.swift#L133-L134

Preparing the reader for writing seeks the reader further if necessary:
https://github.com/dfed/CacheAdvance/blob/0e63e448850a0d64a2a228d59d293749c3c1a248/Sources/CacheAdvance/CacheAdvance.swift#L209-L222

Once we've done that, we pull out the current reader's `offsetInFile` and assign that value to a local `offsetInFileOfOldestMessage` property:
https://github.com/dfed/CacheAdvance/blob/0e63e448850a0d64a2a228d59d293749c3c1a248/Sources/CacheAdvance/CacheAdvance.swift#L136-L137

So in conclusion, what we're doing is reseting the `offsetInFileOfOldestMessage` to `expectedEndOfHeaderInFile`, then seeking `offsetInFileOfOldestMessage` forward significantly, and _then_ syncing the new `offsetInFileOfOldestMessage` to the header.

Reply:
Re: https://github.com/dfed/CacheAdvance/pull/64#discussion_r992451299 the `end of newest` should point to the end of `message7` not the start, right? I want to make sure I'm not missing something.
<img width="832" alt="Screen Shot 2022-10-11 at 6 23 17 PM" src="https://user-images.githubusercontent.com/1791049/195227477-7969d3a5-3273-4268-a0e0-bdb478bf1795.png">

Reply:
You're right @bachand. I've updated the comment. Appreciate the close read!